### PR TITLE
Add the new lint `same_item_push`

### LINKED ISSUE #4078: Suggest `resize` rather when pushing same item to vec
Example in https://github.com/Gymmasssorla/finshir/blob/eae5d0c6761f5558c8ed33ef098d8bd13a07e64f/src/testing/helpers.rs#L55-L61

When pushing the same item to vec.

```rust
fn gen_portions() -> Vec<Vec<u8>> {
    let mut spaces = Vec::with_capacity(EMPTY_SPACES_COUNT);
    for _ in 0..EMPTY_SPACES_COUNT {
        spaces.push(vec![b' ']);
    }
    spaces
}
```

Could be.

```rust
fn gen_portions() -> Vec<Vec<u8>> {
    let mut spaces = Vec::with_capacity(EMPTY_SPACES_COUNT);
    spaces.resize(EMPTY_SPACES_COUNT, vec![b' ']);
    spaces
}
```

Maybe we could even use `vec![b''].repeat(EMPTY_SPACES_COUNT)` after stabilization of`repeat_generic_size`? https://github.com/rust-lang/rust/issues/48784

### PR DESCRIPTION
**Pull request by giraffate**
changelog: Add the new lint `same_item_push`

Fixed #4078. As I said in https://github.com/rust-lang/rust-clippy/issues/4078#issuecomment-658184195, I referrerd to https://github.com/rust-lang/rust-clippy/pull/4647.

### CODE REVIEW

**Comment by giraffate** [comment]
I couldn't run rustfmt at nightly so I haven't run rustfmt yet.
```
$ rustc -V
rustc 1.47.0-nightly (d7f945163 2020-07-19)
$ cargo dev fmt                 Finished dev [unoptimized + debuginfo] target(s) in 0.08s
     Running `clippy_dev/target/debug/clippy_dev fmt`
error: A command failed! `cd /Users/tnakata/workspace/rust-clippy && rustfmt +nightly --version`
$ rustfmt --version
error: the 'rustfmt' component which provides the command 'rustfmt' is not available for the 'nightly-x86_64-apple-darwin' toolchain
```

**Comment by giraffate** [comment]
I run rustfmt in a little old nightly.
```
$ rustc -V
rustc 1.46.0-nightly (346aec9b0 2020-07-11)
```

File: clippy_lints/src/loops.rs [hunk 1]
```diff
@@ -1017,6 +1052,238 @@ fn detect_manual_memcpy<'tcx>(
     }
 }
 
+// Delegate that traverses expression and detects mutable variables being used
+struct UsesMutableDelegate {
```
Comment (Manishearth) [inline clippy_lints/src/loops.rs:1056]: Can this stuff be in `utils`?

File: clippy_lints/src/loops.rs [hunk 2]
```diff
@@ -1017,6 +1052,238 @@ fn detect_manual_memcpy<'tcx>(
...
+            cx.typeck_results(),
+        )
+        .walk_expr(expr);
+    });
+
+    delegate.found_mutable
+}
+
+// Scans for the usage of the for loop pattern
+struct ForPatternVisitor<'a, 'tcx> {
```
Comment (Manishearth) [inline clippy_lints/src/loops.rs:1092]: This seems unnecessary, we just need to walk and look for a path that overlaps, we don't need to do any highly bespoke visiting for that.

Comment (Manishearth) [reply clippy_lints/src/loops.rs:1092]: I think it's actually better to just check if the for pattern contains any new bindings (i.e. it only contains `_` and constants). Unused bindings will be warned about by the unused variable lint

File: clippy_lints/src/loops.rs (hunk 1 above)
Comment (giraffate) [reply clippy_lints/src/loops.rs]: I used [`mutated_variables`](https://github.com/rust-lang/rust-clippy/blob/5a20489c5ca0951827cbb1b4d72ddfdfd393713a/clippy_lints/src/utils/usage.rs#L14) in utils. How about this?

File: clippy_lints/src/loops.rs (hunk 2 above)
Comment (giraffate) [reply clippy_lints/src/loops.rs]: Thanks for your review! I fixed to just check if `for pattern` contains `_` in https://github.com/rust-lang/rust-clippy/pull/5825/commits/5670bc6149027c308abdf2feb708aef61e84943b. Is my understanding correct?

**Comment by giraffate** [comment]
Test failed where it seems unrelated.

**Comment by giraffate** [comment]
@Manishearth Thanks for your review! I fixed it and tests passed now, so could you take another look into this?

**Pull request merged**